Automattic/amp-wp
# Implement dependency resolution for services

## Summary


Fixes #6459

This PR implements a simple way for services to resolve their dependent services so that it is guaranteed that those services exist before retrieving them from the service container.

In order to accomplish this, the `AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin::register_services()` method has been modified to check if a service implements the `AmpProject\AmpWP\Infrastructure\HasRequirements` interface, and if it does, it then retrieves the list of service identifiers it requires in order for the service to be registered. If any of the required service IDs do not exist in the service container, the registration of the service will be delayed (moved to the end of the to-be-registered services list) to allow for other services to be registered in the mean time.

## Testing

To test that this fixes #6459:

1. Add the following plugin to replicate the bug:

```php
<?php
/**
 * Plugin Name: Activate AMP Plugin
 */

add_action( 'admin_head', function () {
    activate_plugin( 'amp/amp.php' );
} );
```

2. Deactivate the AMP plugin
3. Activate the plugin above and it should successfully activate the AMP plugin without any errors.

## Checklist
